x-pack/plugin/esql-core/src/main/java/org/elasticsearch/xpack/esql/core/expression/Attribute.java

```diff
@@ -149,4 +149,9 @@ public static boolean dataTypeEquals(List<Attribute> left, List<Attribute> right
         }
         return true;
     }
+
+    /**
+     * @return true if the attribute represents a TSDB dimension type
+     */
+    public abstract boolean isDimension();
```

Comment:
I wonder if this should actually be ternary, resp. return `Boolean` (capital `B`) so that attributes can return `null` to say "I don't know if I refer to a dimension" rather than claiming not to.

Otherwise :+1:

Reply:
What would be the behavior associated with the null version?  The intention here is that we want to only allow filtering by dimensions, and enable automatic grouping by all dimensions.  If a field returns null here, we would then need to decide if "null" behaved as "this is a dimension" or "this is not a dimension" for these cases.

Unless you're proposing that we fail (or null-with-warning) any query where there is ambiguity about the dimensions?  I'm not sure I agree with that.